packages/http/src/validator/validators/body.ts

```diff
@@ -1,82 +1,131 @@
+      Either.chain(partial(validateBody, schema!)),
+      Either.fold(partial(applyPrefix, this.prefix), () => []),
+    );
+  }
+}
 
-        if (encoding.style) {
-          const deserializer = body.get(encoding.style);
-          if (deserializer && schema.properties) {
-            const propertySchema = schema.properties[property];
-            target[property] = deserializer.deserialize(property, target, propertySchema as JSONSchema);
-          }
-        }
-      }
-    }
+function validateBody(schema: JSONSchema, target: any) {
+  const diagnostics = validateAgainstSchema(target, schema);
+  return diagnostics.length ? Either.left(diagnostics) : Either.right(target);
+}
+
+function maybeValidateFormBody(
+  schema: JSONSchema,
+  content: IMediaTypeContent,
+  mediaType: string | undefined,
+  target: any,
+) {
+  if (mediaType && typeIs.is(mediaType, 'application/x-www-form-urlencoded')) {
+    const encodings = get(content, 'encodings', []);
+    const encodedUriParams = splitUriParams(target);
 
-    return validateAgainstSchema(target, schema).map(error =>
-      Object.assign({}, error, { path: [prefix, ...(error.path || [])] }),
+    return pipe(
+      validateAgainstReservedCharacters(encodedUriParams, encodings),
+      Either.map(decodeUriEntities),
+      Either.map(deserializeFormBody.bind(undefined, schema, encodings)),
     );
   }
 
-  private getContent(specs: IMediaTypeContent[], mediaType?: string): IMediaTypeContent | undefined {
-    if (!mediaType) {
-      return specs[0];
-    }
+  return Either.right(target);
+}
+
+function applyPrefix(prefix: string, diagnostics: IPrismDiagnostic[]): IPrismDiagnostic[] {
+  return diagnostics.map(d => ({ ...d, path: [prefix, ...(d.path || [])] }));
+}
+
+function validateAgainstReservedCharacters(encodedUriParams: Dictionary<string, string>, encodings: IHttpEncoding[]) {
+  return pipe(
+    encodings,
+    Array.reduce<IHttpEncoding, IPrismDiagnostic[]>([], (diagnostics, encoding) => {
+      const allowReserved = get(encoding, 'allowReserved', false);
+      const property = encoding.property;
+      const value = encodedUriParams[property];
 
-    const content = specs.find(c => c.mediaType === mediaType);
+      if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
+        diagnostics.push({
+          path: [property],
+          message: 'Reserved characters used in request body',
+          severity: DiagnosticSeverity.Error,
+        });
+      }
 
-    if (!content) {
-      return specs[0];
-    }
+      return diagnostics;
+    }),
+    diagnostics => (diagnostics.length ? Either.left(diagnostics) : Either.right(encodedUriParams)),
+  );
+}
 
-    return content;
+function deserializeFormBody(
+  schema: JSONSchema,
+  encodings: IHttpEncoding[],
+  decodedUriParams: Dictionary<string, string>,
+) {
+  if (!schema.properties) {
+    return decodedUriParams;
   }
+
+  return pipe(
```

Comment:
Is `pipe` really required here?

Reply:
I could achieve the same using old good `Object.keys().reduce()`. I'm unsure which to pick TBH. Personally I'm always fan of producing short code. On the other hand we are trying to be as FP as possible, so that's why I've chosen pipe. Any vote will tip the balance on any approach.

Reply:
If the choice about `reduce()` vs `pipe()`, then none of these is more FP than the other